core/node.go

```diff
@@ -143,9 +198,51 @@ func (n *Node) ID() peer.ID {
 // Start causes the Node to continuously send messages to and receive messages
 // from its peers. It blocks until an error is encountered or `Stop` is called.
 func (n *Node) Start() error {
+	// Note: dht.Bootstrap has a somewhat confusing name. It doesn't automatically
+	// connect to the bootstrap peers. It just starts the background process of
+	// searching for new peers.
+	if err := n.dht.Bootstrap(n.ctx); err != nil {
+		return err
+	}
+
+	// Advertise ourselves for the purposes of peer discovery.
+	discovery.Advertise(n.ctx, n.routingDiscovery, n.config.RendezvousString, discovery.TTL(advertiseTTL))
+
+	if n.config.UseBootstrapList {
+		if err := n.connectToBootstrapList(); err != nil {
+			return err
+		}
+	}
+
 	return n.mainLoop()
 }
 
+func (n *Node) connectToBootstrapList() error {
+	log.WithField("bootstrapPeers", dht.DefaultBootstrapPeers).Info("connecting to bootstrap peers")
+	connectCtx, cancel := context.WithTimeout(n.ctx, defaultNetworkTimeout)
+	defer cancel()
+	wg := sync.WaitGroup{}
+	// TODO(albrow): Replace default list with our own bootstrap peer list.
+	for _, addr := range dht.DefaultBootstrapPeers {
+		peerInfo, err := peerstore.InfoFromP2pAddr(addr)
+		if err != nil {
+			return err
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if err := n.host.Connect(connectCtx, *peerInfo); err != nil {
```

Comment:
`Connect` will establish a connection and trigger the notifees on protocols like the DHT. Those protocols will in turn open streams to the connected peer to test whether the other party supports the protocol, and will perform any init logic in consequence (e.g. adding the peer to a routing table). Until that happens, you won't have a working DHT node. Since the notifees are triggered asynchronously, so you should add like a 2 second wait after all dials completed.

We are speccing out a local event bus (will tag you in that issue) that will allow us to be deterministic instead of probabilistic in the near future (you'll listen to an event instead of relying on a timing condition), but until that's a thing, this is probably the safest path.

Reply:
Okay I'll add a 2 second wait after dialing all the bootstrap peers for now. Also wanted to point out that the chat-with-rendezvous example doesn't do this: https://github.com/libp2p/go-libp2p-examples/blob/master/chat-with-rendezvous/chat.go. Should we add a `time.Sleep` there as well?